expr/plan.go

```diff
@@ -0,0 +1,246 @@
+			}
+		}
+		j += 1
+	}
+	// when we stop iterating, j should be a non-existant pos
+	if len(e.args) > j {
+		return nil, ErrTooManyArg
+	}
+	err := fn.Init(e.args)
+	if err != nil {
+		return nil, err
+	}
+	from, to = fn.Depends(from, to)
+	// look at which arguments are requested
+	// if the args are series, they are to be requested with the potentially extended to/from
+	// if they are not, keep traversing the tree until we find out which metrics to fetch and for which durations
+	for _, arg := range e.args {
+		if arg.etype == etName || arg.etype == etFunc {
+			reqs, err = newplan(arg, from, to, stable, reqs)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return reqs, nil
+}
+
+// Run invokes all processing as specified in the plan (expressions, from/to) with the input as input
+func (p Plan) Run(input map[Req][]models.Series) ([]models.Series, error) {
+	var out []models.Series
+	p.input = input
+	p.generated = make(map[Req][]models.Series)
+	for _, expr := range p.exprs {
+		o, err := p.run(p.From, p.To, expr)
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, o...)
+	}
+	return out, nil
+}
+
+func (p Plan) run(from, to uint32, e *expr) ([]models.Series, error) {
+	if e.etype != etFunc && e.etype != etName {
+		panic("this should never happen. request must be a function call or metric pattern")
+	}
+	if e.etype == etName {
+		req := Req{
+			e.target,
+			from,
+			to,
+		}
+		return p.input[req], nil
+	}
+
+	// here e.type is guaranteed to be etFunc
+	fdef, ok := funcs[e.target]
+	if !ok {
+		panic(fmt.Sprintf("cannot find func %q. this should never happen as we should have validated function existence earlier", e.target))
+	}
+	fn := fdef.constr()
+	err := fn.Init(e.args)
+	if err != nil {
+		return nil, err
+	}
+	from, to = fn.Depends(from, to)
+	// look at which arguments are requested
+	// if the args are series, they are to be requested with the potentially extended to/from
+	// if they are not, keep traversing the tree until we find out which metrics to fetch and for which durations
+	results := make([]interface{}, len(e.args))
+	for i, arg := range e.args {
+		if arg.etype == etName || arg.etype == etFunc {
+			result, err := p.run(from, to, arg)
+			if err != nil {
+				return nil, err
+			}
+			results[i] = result
+		} else if arg.etype == etString {
+			results[i] = arg.valStr
+		} else {
+			// etype == etConst
+			results[i] = arg.val
+		}
+	}
+	// we now have all our args and can process the data and return
+	rets, err := fn.Exec(p.generated, results...)
+	if err != nil {
+		return nil, err
+	}
+	series := make([]models.Series, len(rets))
+	for i, ret := range rets {
+		series[i] = ret.(models.Series)
+	}
+	return series, nil
+}
+
+// Clean returns all buffers (all input data + generated series along the way)
+// back to the pool.
+func (p Plan) Clean() {
+	for _, series := range p.input {
+		for _, serie := range series {
+			pointSlicePool.Put(serie.Datapoints[:0])
```

Comment:
it creates an empty list by creating a slice value, backed by (using a pointer to) the same underlying array. so when we pull the value out of the pool next time, we have a slice of len zero that we can immediately start adding elements to without allocations, because we'll reuse the same backing array.